### DISCUSSION TITLE
Testing All Features

### DISCUSSION THREAD

**Discussion by Sarveshgithub** [Sarveshgithub/sfdc-lwc-lightning-datatable#71] [General]
We have released some features in the past weeks, we have to do some regression testing.

1. Check with different types of field data types ( Date, Email, Boolean, DateTime, lookup )
2. Edit - check for Related, Id fields ( Edit not allow ), edit with different types of fields
4. Pagination functionality
5. Show All/ Hide All
6. Is Pagination + Show All will work together?
7. Search - check search functionality with pagination, is pagination breaking with search, and search works with all fields which are present in fields

@mickaelgudin The above feature I have listed, we have to verify, feel free to add if I miss some test cases.

**Comment by mickaelgudin** [comment]
You're right, we need to test everything.
We should comment on this discussion every cases tested and potential bug/enhancement here

With the following template :

**What I have tested :**

**Input :**

**Result :**

**Reply by Sarveshgithub** [reply]
@mickaelgudin  can I assign this to you? , If any issue come, will document the issue and create issues.

**Comment by mickaelgudin** [comment]
**DOCUMENTATION**

**What I have tested :** Have the list without any icon (possible need)

**Input :**  I entered "j" in the icon name (which is not a valid icon name)

**Result :** we should add this in the readme (enter "disabled" in the icon name to remove the icon)

**Reply by Sarveshgithub** [reply]
@mickaelgudin yhea for this one, the user should enter a valid [icon](https://www.lightningdesignsystem.com/icons), we will provide slds icon [link](https://www.lightningdesignsystem.com/icons/), documentation [here](https://github.com/Sarveshgithub/sfdc-lwc-lightning-datatable#steps-to-customization-through-design-attribute) in `Example` column

priority - Low

**Comment by mickaelgudin** [comment]
**Implementation / enhancement** #22 

**What I have tested :** Edit a record inline

**Input :** change the Description field of a contact

**Result :** 

The redirection doesn't work after save

I believe we should have a new prop "enable/disable redirection on save" :

if disabled the record would be updated(fetchRecord) after save

**Reply by Sarveshgithub** [reply]
@Sarveshgithub can you please tell the design values, In below format
```
// cmp.iconName = "standard:account";
// cmp.recordId = "recid";
// cmp.relatedFieldAPI = "AccountId";
// cmp.whereClause;
cmp.title = "Contacts";
cmp.fields = "FirstName";
cmp.objectName = "Contact";
cmp.limit = 5;
cmp.isCounterDisplayed = true;
cmp.actionButtons = '[{"name": "New","label": "New","variant": "neutral"}]';
cmp.showCheckboxes = true;
cmp.showViewAll = false;
cmp.hasPagination = false;
cmp.hasSearchBar = false;
/*cmp.predefinedCol =
	'{"Account.Name":{"label":"Account Name","type":"url","typeAttributes":{"label": { "fieldName": "Account.Name" ,"recId": "AccountId"}}},"AccountId":{"label":"Ac Id","type":"Id"}}';
*/
```

**Reply by mickaelgudin** [reply]
@Sarveshgithub 

The problem with the edit line is that **the redirection is made based on this.recordId** but the component can be used elsewhere(in my test it was in the homepage).

**Instead of a redirection maybe a fetch of records could be great, to avoid refreshing the page even if we are in a record page.**

**Reply by Sarveshgithub** [reply]
@mickaelgudin Can you please tell me which redirection, is this Row action(View) ? or Row level record redirection? can you share a screenshot please?

**Reply by Sarveshgithub** [reply]
@mickaelgudin Are you refring to this one?
<img width="972" alt="image" src="https://user-images.githubusercontent.com/40709233/189629227-0194a2cd-16b1-4732-83d8-284bf25615f0.png">

**Reply by mickaelgudin** [reply]
@Sarveshgithub yes that's it

It doesn't work on pages that are not a record page

**Reply by Sarveshgithub** [reply]
@mickaelgudin Ahh good catch, let me create an Issue, I will take a look, thanks 😀

**Reply by Sarveshgithub** [reply]
The issue has been [fixed](https://github.com/Sarveshgithub/sfdc-lwc-lightning-datatable/issues/91), @mickaelgudin can you please re check from `dev`branch.

**Reply by mickaelgudin** [reply]
@Sarveshgithub The fix is fine since there is no error. 

However, if there is no recordId 

we should either : 

- call fetchRecords because the old values are kept after saving(even if the update succeeded).
- redirect to the current page

**Reply by Sarveshgithub** [reply]
Agree, fetchRecords() is a good option, we get rid of from showDetail(). but we need to verify pagination as well, i guess it will work. let me change to fetchRecords()

**Reply by Sarveshgithub** [reply]
done with changes

**Comment by mickaelgudin** [comment]
**type :** bug

**feature :** Pagination


**What I have tested :** Pagination without a limit(all records)

**Input :** all contacts name without limit

**Result :** the "next" and "last" buttons are not disabled

**Reply by Sarveshgithub** [reply]
Created Issue - https://github.com/Sarveshgithub/sfdc-lwc-lightning-datatable/issues/93

**Reply by Sarveshgithub** [reply]
FIxed in 'dev' branch

**Comment by mickaelgudin** [comment]
**type :** bug

**feature :** Pagination


**What I have tested :** Pagination "last" button

**Input :** all contacts name with a limit of 10 and with 20 contacts in total

**Result :** the "last" button is not working

**cause :** method lastPage in the javascript controller

the current 'formula' is `this.offSet = (this.totalRows / this.limit) * this.limit`

I believe it should be : `this.offSet = this.totalRows - this.limit`

**Reply by Sarveshgithub** [reply]
Created Issue : https://github.com/Sarveshgithub/sfdc-lwc-lightning-datatable/issues/94

**Reply by Sarveshgithub** [reply]
FIxed in 'dev' branch

**Comment by mickaelgudin** [comment]
**Topic : Is Pagination + Show All will work together?**

The current behavior with the two features enabled is ok, I didn't find any bugs. 

There might be uses cases using both we can leave it as is just in case it's useful.

**Reply by Sarveshgithub** [reply]
@mickaelgudin Yhea sounds good.

**Comment by mickaelgudin** [comment]
**feature :** search

**What I have tested :** Search with a where

**Input :** all contacts but excluding one

**Result :** the contact not displayed in the datatable is shown on search

**cause :** the apex method onSearch doesn't take into account the where clause


I don't know if we should allow admins to configure the component to allow "global" search (current behavior). But I know most use cases don't need the current behavior and we expect the where clause to be taken into account.

**Reply by Sarveshgithub** [reply]
@mickaelgudin hmm, can we append where in SOSL, if where!= null

**Reply by mickaelgudin** [reply]
yes no need for another issue, we can consider this as part of #30

**Reply by Sarveshgithub** [reply]
done

**Comment by mickaelgudin** [comment]
**feature :** sorting

**What I have tested :** Sorting with all columns orders

**Input :** Sort several times on several fields

**Result :** The arrow are not always in the good order or rerendering

**cause :** 
in the template add `sorted-direction={sortDirection}` in as a prop of lightning-datatable

**Reply by Sarveshgithub** [reply]
@mickaelgudin This is Bug

**Reply by Sarveshgithub** [reply]
done

**Comment by mickaelgudin** [comment]
**Topic :** configure the component to be available in community

we should add in meta file : `<target>lightningCommunity__Page</target>`

**Reply by Sarveshgithub** [reply]
@mickaelgudin This is Enhancement

**Reply by Sarveshgithub** [reply]
done

**Comment by mickaelgudin** [comment]
**feature :** Search with pagination enabled

**What I have tested :** Search then click on button "previous" or "next" or "first" or "last"

**Input :** Input with result in next page or in previous page

**Result :**  When we click on "Next" or "Previous" the offset before the search is still active.


**Proposed enhancement :**

- if there is a search term then the pagination buttons are disabled(because all search results are displayed on search(there is no limit in the SOSL query) and I believe it should remain that way)
- replace the type of the lightning-input (it should be type="search" not "text", because the search type provides a search icon + a cross to remove entry)

**Reply by Sarveshgithub** [reply]
done

**Comment by mickaelgudin** [comment]
**DOCUMENTATION**

**error :** @Sarveshgithub the customized json provided in the readme doesn't work (we need to replace single quote into double quote)

same goes for buttons configuration (only double quote work)

**Reply by Sarveshgithub** [reply]
yhea will update the doc

**Comment by mickaelgudin** [comment]
**feature :** inline edit

@Sarveshgithub The boolean / checkbox case works fine

**Comment by mickaelgudin** [comment]
feature : inline edit

@Sarveshgithub currency fields edition works fine

**Comment by mickaelgudin** [comment]
@Sarveshgithub Hey,

your recent commit on the master branch to format code has an issue :

please use double quote instead of single quote(as before) on the readme.md for json examples otherwise an error is throwned

**Comment by Sarveshgithub** [comment]
@mickaelgudin yhea let me skip the readme file on format script

**Comment by mickaelgudin** [comment]
@Sarveshgithub 

Could you add the following example to the readme(in the dev branch and master) in label override section ?  (**right after multiple override**), I tested the provided example and it's working just fine :


#### Label Override using a label (for translations the connected user language is taken into account)

```yml
{ "FirstName": { "label": "{!Label.MyLabelName}", "type": "text" } }
```

**Reply by Sarveshgithub** [reply]
yup added

**Comment by mickaelgudin** [comment]
@Sarveshgithub could you create a new release and mention the new features : 

- JSON not required for basic fields (new prop Customized Field JSON, only for lookup, picklist, record redirection and label override)
- search bar
- sort
- order by clause
- picklist edit
- lookup edit

**Comment by Sarveshgithub** [comment]
Hi @mickaelgudin How you doing, i just optimized Apex class, i did some Unit test, didn't find any issue, can you also verify from your end. is it working everything fine. latest in dev branch

**Reply by mickaelgudin** [reply]
Hi @Sarveshgithub, no problem, I'll look into it. When I'm done I'll give you a feedback.

**Reply by mickaelgudin** [reply]
Hello @Sarveshgithub, I reviewed the changes they seemed good (it provides more readable apex classes).


However, I believe lookups field are broken when we only want to have a lookup with a link to the record :

You can take the default values(contact with a lookup that links to the account) when putting the component in a page layout through app builder.


You'll see that there's an error : 

![image](https://user-images.githubusercontent.com/39730173/236677687-fdc93bf2-f82d-41e0-96a7-ad19aa91473d.png)

I believe this error has something to do with one of your changes since the code in the master branch do not throw this error. I'll let you dig into it.

**Reply by Sarveshgithub** [reply]
@mickaelgudin fixed this try to retest

**Reply by mickaelgudin** [reply]
@Sarveshgithub I've looked at your recent changes on the IconUtils, it will only work on standard objects or custom object with standard icon.

Can you do a rollback of the IconUtils class on the dev branch?

I'll tests everything when you're done before you merge it to master.

**Reply by Sarveshgithub** [reply]
@mickaelgudin IconUtils changes are intentional because the custom icons code was not covered in the test class and deployment is failing, now it's all 100% overall apex code, I made the custom icon name as static

**Reply by mickaelgudin** [reply]
@Sarveshgithub I'll look into it and give you a feedback

**Reply by mickaelgudin** [reply]
@Sarveshgithub As promised I'm starting to test and review your changes on the dev branch,
I'll do a feedback soon.

**Reply by mickaelgudin** [reply]
@Sarveshgithub :

- regression on custom icon : The custom icons are not retrieved (it's the icon custom7 that is displayed in this case)
- regression on picklist values : picklist values are not retrieved when we edit a picklist field if we do not specify that the column is a picklist in the custom json (please uncomment the section below)

![image](https://github.com/Sarveshgithub/sfdc-lwc-lightning-datatable/assets/39730173/4651d864-6520-43f5-b23a-51d3704192df)


There might be other way to enhance coverage (less new lines for the uncover code in the IconUtils class) without causing the above regression on the sobject icon autoretrieve feature.

**Comment by mickaelgudin** [comment]
@Sarveshgithub could you merge the master branch into the dev branch?

**Reply by Sarveshgithub** [reply]
done

**Comment by RemyGenius** [comment]
Hi @Sarveshgithub ,

Everything works well in sandbox but when I tried to deploy to Production, I get an error on the APEX Test Classe : 

Deployment CompleteTest failure, method: RelatedList_Test.buildFieldJSON_Test -- System.AuraHandledException: Script-thrown exception stack Class.RelatedList.buildFieldJSON: line 183, column 1
Class.RelatedList_Test.buildFieldJSON_Test: line 19, column 1

Could you please help me to fix this issue?

Thank you,

**Comment by mickaelgudin** [comment]
Hi @Sarveshgithub, 
hope you're doing well, could you correct this typo error in the repo description :

![image](https://github.com/Sarveshgithub/sfdc-lwc-lightning-datatable/assets/39730173/275566dc-ad27-4098-b93b-f1c8b86b1020)

**Reply by Sarveshgithub** [reply]
@mickaelgudin Yhea all good, thanks for notice, fixed it

**Comment by mickaelgudin** [comment]
Hi @Sarveshgithub hope you're doing ok,
can you change the repo description from 'Salesforce Lighting Data table' to 'Salesforce Lightning Datatable' ?

**Reply by Sarveshgithub** [reply]
@mickaelgudin yhea doing good, changed it
